File: RotateArray.kt
// https://leetcode.com/problems/rotate-array/


class Solution {
    fun rotate(nums: IntArray, k: Int): Unit {
        val originalArray = nums.copyOf()
        for (i in nums.indices) {
            val j = (i + k) % nums.size
            nums[j] = originalArray[i]
        }
    }
}


/* 
Time Complexity -> O(n)
Space Complexity -> O(n) 
 */

//-----------------------------------------------------------------------

/* 
Time Complexity -> O(n)
Space Complexity -> O(1) 
 */
class Solution {
    fun dorotate(nums:IntArray, low: Int, high: Int): Unit {
        var a = low
        var b = high
        while(a < b){
            nums[a] = nums[b].also{nums[b] = nums[a]}
            a++
            b--
        }
    }
    
    fun rotate(nums: IntArray, k: Int): Unit {
            val tempk = k % nums.size
            val n = nums.size
            dorotate(nums, n - tempk, n - 1)
            dorotate(nums, 0, n - tempk - 1)
            dorotate(nums, 0, n - 1)
                
        }
    }

File: RotateArray.py
# https://leetcode.com/problems/rotate-array/

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def dorotate(nums, low, high):
            a = low
            b = high
        
            while(a < b):
                nums[a], nums[b] = nums[b], nums[a]
                a = a + 1
                b = b - 1
            
            
        n = len(nums)
        k = k % n
        dorotate(nums, n - k, n - 1)
        dorotate(nums, 0, n - k - 1)
        dorotate(nums, 0, n - 1)
        
  '''
  Time complexity -> O(n)
  Space Complexity -> O(1)
  '''
